### PR TITLE
Replace bindings by WBT blocks through Simulink function calls

### PR DESCRIPTION
**Pull request by nunoguedelha**
Implements #5 and subsequent fixes (refer to https://github.com/dic-iit/matlab-whole-body-simulators/issues/5#issuecomment-771047600).

Default iCub simulation should work exactly as before (should render same robot behavior), except that it should run approximately 3 times faster.

### CODE REVIEW

**Comment by nunoguedelha** [comment]
Ping @GiulioRomualdi , @Giulero .

**Comment by nunoguedelha** [comment]
For the last commit, refer to https://github.com/dic-iit/matlab-whole-body-simulators/issues/5#issuecomment-772948255.

**Comment by Giulero** [comment]
Hi @nunoguedelha ! Maybe I've something wrong in my configuration, but I've got this error
```matlab
An error occurred in the block 'test_matlab_system_19/RobotDynWithContacts/MATLAB System' during compile.
Caused by:
Simulink detected an error 
 'Simulink encountered an error while allocating a handle object in MATLAB System block 'test_matlab_system_19/RobotDynWithContacts/MATLAB System'. Create the handle object only once in the corresponding System object 'robotDynWC.step_block'.'.

 The error occurred for MATLAB System block 'test_matlab_system_19/RobotDynWithContacts/MATLAB System'.    

 To prevent this error, use one of the following:
 * Modify the System object to avoid code that does not support code generation.
 * Change 'Simulate using' parameter to 'Interpreted Execution'. 
 ```

File: init.m
Comment (Giulero) [inline init.m:46]: Does it make sense to add to the path `src` and `lib` here? I had to add them manually

File: lib/+iDynTree2WBTmappers/KinDynComputations.m
Comment (Giulero) [inline lib/+iDynTree2WBTmappers/KinDynComputations.m:6]: Double semicolon

File: lib/+iDynTree2WBTmappers/KinDynComputations.m
Comment (Giulero) [inline lib/+iDynTree2WBTmappers/KinDynComputations.m:8]: Double semicolon

File: lib/+iDynTree2WBTmappers/KinDynComputations.m
Comment (Giulero) [inline lib/+iDynTree2WBTmappers/KinDynComputations.m:37]: Not sure about this. If the error occurs, the program should exit and the ack will be not changed, isn't it?

File: lib/+wbs/@Contacts/Contacts.m
Comment (Giulero) [inline lib/+wbs/@Contacts/Contacts.m:168]: Love your vectorization! :D

File: lib/+wbs/@Robot/Robot.m
Comment (Giulero) [inline lib/+wbs/@Robot/Robot.m:92]: Typo: left -> right

**Comment by nunoguedelha** [comment]
> Hi @nunoguedelha ! Maybe I've something wrong in my configuration, but I've got this error
> 
> ```matlab
> An error occurred in the block 'test_matlab_system_19/RobotDynWithContacts/MATLAB System' during compile.
> Caused by:
> Simulink detected an error 
>  'Simulink encountered an error while allocating a handle object in MATLAB System block 'test_matlab_system_19/RobotDynWithContacts/MATLAB System'. Create the handle object only once in the corresponding System object 'robotDynWC.step_block'.'.
> 
>  The error occurred for MATLAB System block 'test_matlab_system_19/RobotDynWithContacts/MATLAB System'.    
> 
>  To prevent this error, use one of the following:
>  * Modify the System object to avoid code that does not support code generation.
>  * Change 'Simulate using' parameter to 'Interpreted Execution'. 
> ```
> 
> Also, when the simulink model stops the file manager moves to `/lib/RobotDynamicsWithContacts`.

Hi @Giulero , it stops in `RobotDynamicsWithContacts` everytime it stops due to an error, otherwise it stops in the original repo root folder `matlab-whole-body-simulators`. I would say this is a normal behavior.
I never saw this error before. Could you start MATLAB from scratch and try again?

File: init.m
Comment (nunoguedelha) [reply init.m:46]: That shouldn't be necessary if you launch the right version of `test_matlab_system_19.mdl` from the repo root path. When the model is loaded, the `PostLoadFcn` function runs the following commands:
```
% Set path to the utility functions and to WBC library
import wbc.*
addpath(genpath('./src'));
addpath('./lib/RobotDynamicsWithContacts');
```
which adds all the required paths, including those you mentioned. for some reason you're loading the wrong mdl version. Try starting MATLAB from scratch...

File: lib/+iDynTree2WBTmappers/KinDynComputations.m
Comment (nunoguedelha) [reply lib/+iDynTree2WBTmappers/KinDynComputations.m:6]: Fixed in next commit.

File: lib/+iDynTree2WBTmappers/KinDynComputations.m
Comment (nunoguedelha) [reply lib/+iDynTree2WBTmappers/KinDynComputations.m:8]: Fixed in next commit.

File: lib/+iDynTree2WBTmappers/KinDynComputations.m
Comment (nunoguedelha) [reply lib/+iDynTree2WBTmappers/KinDynComputations.m:37]: Absolutely. There was actually a warning. I kept the `ack` variable for conserving the previous interface of the usd `KinDynComputations` methods. I had in mind that that we might want to get back to using the bindings if some day we manage to generate C++ code from that. For that reason I created a `KinDynComputations` class inside a new package `matlab-whole-body-simulators/lib/+iDynTree2WBTmappers`.

For the sake of coherence, I removed the call to `error` in this method, as the `ack` is checked by the caller, which triggers an error on its own.

Comment (nunoguedelha) [reply lib/+iDynTree2WBTmappers/KinDynComputations.m:37]: Fixed in the next commit.

File: lib/+wbs/@Contacts/Contacts.m
Comment (nunoguedelha) [reply lib/+wbs/@Contacts/Contacts.m:168]: :D actually it was the easiest solution to have a code supported by the code generator (dynamic matrices building is not supported), and have a variable `J`.

File: lib/+wbs/@Robot/Robot.m
Comment (nunoguedelha) [reply lib/+wbs/@Robot/Robot.m:92]: bad copy paste :D
Fixed in next commit.

**Comment by nunoguedelha** [comment]
Rework completed. @Giulero

**Comment by nunoguedelha** [comment]
@Giulero , @GiulioRomualdi , I've updated the README for reflecting the Config.USE_OSQP flag deprecation.

**Comment by Giulero** [comment]
I tried starting Matlab from scratch and run `test_matlab_system_19`. 
Same error :(

**Comment by nunoguedelha** [comment]
> I tried starting Matlab from scratch and run `test_matlab_system_19`.
> Same error :(

Hi @Giulero  I'm trying to reproduce it in my environment...

**Comment by nunoguedelha** [comment]
Let's see if @CarlottaSartore (MATLAB 2019b + Linux) and @VenusPasandi (MATLAB2020b + Linux) have the same issue.

I definitely can't reproduce it (MATLAB 2019b + MacOS).

**Comment by CarlottaSartore** [comment]
My computer Configuration: 

- Ubuntu 18.04 LTS
- matlab-whole-body-simulator
~~~
commit e14d45088f4f4ec1896f075311538739b35c70dc (HEAD -> feature/replace-bindings-by-wbt-blocks, origin/feature/replace-bindings-by-wbt-blocks)
Author: Nuno Guedelha <nuno.guedelha@iit.it>
Date:   Thu Feb 4 23:27:04 2021 +0100

    Update the README w.r.t. the deprecation of Config.USE_OSQP and temporary use of qpOASES
~~~
- IDynTree v2.0.2
- WBtoolbox 
 ~~~ commit 0db6ae4c8a5f035d70ad2e1f5c0c8762ff51ad8a (HEAD, origin/devel)
Merge: 9a3848b 44f3d8b
Author: Diego Ferigo <diego.ferigo@iit.it>
Date:   Tue Nov 24 14:23:23 2020 +0100

    Merge pull request #194 from CarlottaSartore/feature/CreateCMMSimulinkBlock
    
    Created CMM Simulink Block
~~~

Able to run ~1 sec of simulation (then the robot falls) and get this error 
~~~

An error occurred while running the simulation and the simulation was terminated
Caused by:
MATLAB System block 'test_matlab_system_19/Robot visualizer/MATLAB System' error occurred when invoking 'stepImpl' method of 'Visualizer'. The error was thrown from '
 '/home/csartore/iit_ws/robotology-superbuild/build/install/mex/+iDynTreeWrappers/updateVisualization.m' at line 21
 '/home/csartore/iit_ws/matlab-whole-body-simulators/src/robotVisualizer/Visualizer.m' at line 42'. 
Invalid value for Matrix property 
~~~

**Comment by VenusPasandi** [comment]
> Let's see if @CarlottaSartore (MATLAB 2019b + Linux) and @VenusPasandi (MATLAB2020b + Linux) have the same issue.

@nunoguedelha , I got the same error as @Giulero said https://github.com/dic-iit/matlab-whole-body-simulators/pull/15#issuecomment-773385872 as following

```
An error occurred in the block 'test_matlab_system_19/RobotDynWithContacts/MATLAB System' during compile.
Caused by:
Simulink detected an error 
 'Simulink encountered an error while allocating a handle object in MATLAB System block 'test_matlab_system_19/RobotDynWithContacts/MATLAB System'. Create the handle object only once in the corresponding System object 'robotDynWC.step_block'.'.

 The error occurred for MATLAB System block 'test_matlab_system_19/RobotDynWithContacts/MATLAB System'.    

 To prevent this error, use one of the following:
 * Modify the System object to avoid code that does not support code generation.
 * Change 'Simulate using' parameter to 'Interpreted Execution'. 
```

**Comment by nunoguedelha** [comment]
Just for clarifying for anyone else joining the discussion:

> @CarlottaSartore : Able to run ~1 sec of simulation (then the robot falls) and get this error

This error is an old error in the visualizer that occurs only when the robot gets to that weird position..

> @VenusPasandi : I got the same error as @Giulero said #15 (comment) as following

this is what we are analyzing.

@VenusPasandi , could you please specify your machine configuration?

Apart from MATLAB version (2020b), do you share the same configuration (superbuild, etc) as @CarlottaSartore ?

**Comment by nunoguedelha** [comment]
After discussing with @VenusPasandi and @CarlottaSartore , we confirm that both the configurations they used are identical except for the MATLAB version. I will upgrade my MATLAB version tomorrow and analyze the problem. Let's wait until Friday at least before we merge this PR.

**Comment by S-Dafarra** [comment]
With Ubuntu 20.04 and Matlab 2020a I had the same error as @CarlottaSartore in https://github.com/dic-iit/matlab-whole-body-simulators/pull/15#issuecomment-775911762.

The robot was kind of exploding. Not sure what is the expected behavior.

**Comment by S-Dafarra** [comment]
Btw, I suggest checking also at the terminal output in which you launched Matlab. In my case I had:
```
sdafarra@iiticublap104:~$ matlab
MATLAB is selecting SOFTWARE OPENGL rendering.
Gtk-Message: 16:10:16.586: Failed to load module "canberra-gtk-module"
Internal qpOASES error. Trying to solve the problem with the remaining number of iterations.
Internal qpOASES error. Trying to solve the problem with the remaining number of iterations.
Internal qpOASES error. Trying to solve the problem with the remaining number of iterations.
Internal qpOASES error. Trying to solve the problem with the remaining number of iterations.
Internal qpOASES error. Trying to solve the problem with the remaining number of iterations.
Internal qpOASES error. Trying to solve the problem with the remaining number of iterations.
Internal qpOASES error. Trying to solve the problem with the remaining number of iterations.
Internal qpOASES error. Trying to solve the problem with the remaining number of iterations.
Internal qpOASES error. Trying to solve the problem with the remaining number of iterations.
Internal qpOASES error. Trying to solve the problem with the remaining number of iterations.
Internal qpOASES error. Trying to solve the problem with the remaining number of iterations.
Internal qpOASES error. Trying to solve the problem with the remaining number of iterations.
Internal qpOASES error. Trying to solve the problem with the remaining number of iterations.
Internal qpOASES error. Trying to solve the problem with the remaining number of iterations.
Internal qpOASES error. Trying to solve the problem with the remaining number of iterations.
Internal qpOASES error. Trying to solve the problem with the remaining number of iterations.
Internal qpOASES error. Trying to solve the problem with the remaining number of iterations.
Internal qpOASES error. Trying to solve the problem with the remaining number of iterations.
Internal qpOASES error. Trying to solve the problem with the remaining number of iterations.
Internal qpOASES error. Trying to solve the problem with the remaining number of iterations.
Internal qpOASES error. Trying to solve the problem with the remaining number of iterations.
Internal qpOASES error. Trying to solve the problem with the remaining number of iterations.
Internal qpOASES error. Trying to solve the problem with the remaining number of iterations.
Internal qpOASES error. Trying to solve the problem with the remaining number of iterations.
Internal qpOASES error. Trying to solve the problem with the remaining number of iterations.
Internal qpOASES error. Trying to solve the problem with the remaining number of iterations.
Internal qpOASES error. Trying to solve the problem with the remaining number of iterations.
Internal qpOASES error. Trying to solve the problem with the remaining number of iterations.
Internal qpOASES error. Trying to solve the problem with the remaining number of iterations.
Internal qpOASES error. Trying to solve the problem with the remaining number of iterations.
Internal qpOASES error. Trying to solve the problem with the remaining number of iterations.
Internal qpOASES error. Trying to solve the problem with the remaining number of iterations.
Internal qpOASES error. Trying to solve the problem with the remaining number of iterations.
Internal qpOASES error. Trying to solve the problem with the remaining number of iterations.
Internal qpOASES error. Trying to solve the problem with the remaining number of iterations.
```
that may explain why the robot was exploding.

**Comment by nunoguedelha** [comment]
> Btw, I suggest checking also the terminal output in which you launched Matlab. In my case I had:

Yes absolutely. At that point the robot was already under the floor a priori. We were having a a similar error way before this PR. Regarding the pqOASES failure, we were planing to handle the error more properly, allowing a given number of consecutive failures while using the last valid QP result, and stopping the simulation if that max number is reached. I've just opened https://github.com/dic-iit/matlab-whole-body-simulators/issues/16 for this.

For clarifying, the robot falling and sinking in the ground is an expected behaviour. The controller used in the test model is a fake controller, we are sending 0 torques to the joints and letting the robot fall on the ground to check the contacts handling.

**Comment by S-Dafarra** [comment]
For the record, this is what I see

https://user-images.githubusercontent.com/18591940/107637932-2e5dc900-6c6f-11eb-9302-b159218dfc4a.mp4

I also noticed that in the Diagnostic Viewer I have the message
> Warning: The new value for the Matrix property may cause rendering problems.
> In iDynTreeWrappers.updateVisualization (line 21)
  In Visualizer/stepImpl (line 42)

It is printed at every visualization update. I confirm that the qpOases errors appear only when the robot is down below the ground.

**Comment by nunoguedelha** [comment]
> For the record, this is what I see

Unfortunately, lately I cannot see videos attached to Github or Gitlab from Linux (probably because of Linux 20?). But I guess it matches my previous description.

Regarding the Warning, I have seen this for a while but as it is not a priori critical nor related to recent developments, I never had time to pay much attention.

**Comment by nunoguedelha** [comment]
After installing MATLAB 2020b, I've reproduced the issue:
image

Analysis ongoing...

**Comment by nunoguedelha** [comment]
## Issue analysis (in Matlab internal code)

Error message `Create the handle object only once in the corresponding System object` found in the xml error definition file `/Applications/MATLAB_R2020b.app/resources/SystemBlock/en/MATLABSystem.xml`, mapping the key `HandleObjectPersistentUnsupported`.

A search on this keyword led to the help page `/Applications/MATLAB_R2020b.app/help/simulink/ug/memory-management-of-handle-objects-in-generated-code.html`.

- We ruled out an issue related to the mask of the library.
- Considering the keyword and the use cases addressed in the help page, we could guess that the issue is related to the section "A Handle Object That a Persistent Variable Refers To Must Be a Singleton Object".
- Some persistent variable might be pointing the object `step_block`, or there is some problem with the "tunable", "nonTunable" parameters of the Matlab System block. **Further analysis ongoing...**

**Comment by nunoguedelha** [comment]
@VenusPasandi @CarlottaSartore @Giulero @S-Dafarra I've edited the analysis https://github.com/dic-iit/matlab-whole-body-simulators/pull/15#issuecomment-778027931.

The fix is working. Pushing it in a few.

**Comment by nunoguedelha** [comment]
Fix pushed and tested. @Giulero, @GiulioRomualdi or @CarlottaSartore , can you please review the fix (very few changes) such that we merge and unblock the other depending PRs?

### Additional info for the review

- main functional hange is
> Removed parent class "handle" from KinDynComputations class definition 
- because of this, the call to a class method changing the class object has to return the updated object (change in `Robot.m`).
- This fix allowed to have the property `robot_config` of "step_block" as Nontunable again.
- Added also `robot_config` to the library mask (minor change for coherence, no functional impact except that this parameter appears clearly as an input of the library, as it was already the case for `contact_config` and `physics.config`).

**Comment by nunoguedelha** [comment]
Hi @GiulioRomualdi , @Giulero , any news on the approval? Sorry to insist, it's just that this PR is critical for improving the simulator performance in another project depending on it, specially this week. CC @CarlottaSartore

**Review by Giulero — APPROVED** [review]

**Comment by nunoguedelha** [comment]
Thx @Giulero ! Merging

**Pull request merged**
